File: src/main/java/com/github/kpavlov/jreactive8583/iso/MessageFactory.java
package com.github.kpavlov.jreactive8583.iso;

import javax.annotation.Nonnull;
import java.io.UnsupportedEncodingException;
import java.text.ParseException;

public interface MessageFactory<T> {

    T newMessage(int type);

    T newMessage(@Nonnull MessageClass messageClass,
                 @Nonnull MessageFunction messageFunction,
                 @Nonnull MessageOrigin messageOrigin);

    T createResponse(T requestMessage);

    T createResponse(T request, boolean copyAllFields);

    T parseMessage(byte[] buf, int isoHeaderLength, boolean binaryIsoHeader) throws ParseException, UnsupportedEncodingException;

    T parseMessage(byte[] buf, int isoHeaderLength) throws UnsupportedEncodingException, ParseException;
}

File: src/main/java/com/github/kpavlov/jreactive8583/iso/MessageFactory.kt
@file:JvmName("MessageFactory")

package com.github.kpavlov.jreactive8583.iso

import java.io.UnsupportedEncodingException
import java.text.ParseException

interface MessageFactory<T> {
    fun newMessage(type: Int): T

    fun newMessage(
        messageClass: MessageClass,
        messageFunction: MessageFunction,
        messageOrigin: MessageOrigin
    ): T

    fun createResponse(requestMessage: T): T
    fun createResponse(request: T, copyAllFields: Boolean): T

    @Throws(ParseException::class, UnsupportedEncodingException::class)
    fun parseMessage(buf: ByteArray, isoHeaderLength: Int, binaryIsoHeader: Boolean): T

    @Throws(UnsupportedEncodingException::class, ParseException::class)
    fun parseMessage(buf: ByteArray, isoHeaderLength: Int): T
}
